include/graphblas/reference/blas3.hpp

```diff
@@ -1282,23 +1743,27 @@ namespace grb {
 		Matrix< OutputType, reference, RIT1, CIT1, NIT1 > &C,
 		const Matrix< InputType1, reference, RIT2, CIT2, NIT2 > &A,
 		const Matrix< InputType2, reference, RIT3, CIT3, NIT3 > &B,
-		const Operator &mulOp,
-		const Phase phase = EXECUTE,
-		const typename std::enable_if< !grb::is_object< OutputType >::value &&
-			!grb::is_object< InputType1 >::value &&
-			!grb::is_object< InputType2 >::value &&
-			grb::is_operator< Operator >::value,
-		void >::type * const = nullptr
+		const Operator &op,
+		const Phase &phase = EXECUTE,
+		const typename std::enable_if<
+				!grb::is_object< OutputType >::value &&
+				!grb::is_object< InputType1 >::value &&
+				!grb::is_object< InputType2 >::value &&
+				grb::is_operator< Operator >::value,
+				void
+			>::type * const = nullptr
 	) {
+		typedef typename std::conditional<std::is_void<InputType1>::value, typename Operator::D1, InputType1>::type ActualInputType1;
+		typedef typename std::conditional<std::is_void<InputType2>::value, typename Operator::D2, InputType1>::type ActualInputType2;
 		// static checks
 		NO_CAST_ASSERT( ( !( descr & descriptors::no_casting ) ||
-			std::is_same< typename Operator::D1, InputType1 >::value ),
+			std::is_same< typename Operator::D1, ActualInputType1 >::value ),
```

Comment:
I'm not sure what is being fixed here -- is it now not the case that `void`  A passes no_casting even if the monoid left-hand input is not void? (Would that not be unexpected behaviour?)

Reply:
It's just to select which of the types will be used: 
* in a void case it will be the operator input type
* in a non-void case it will be the one of the matrix